Add ANSI-X9.63-KDF

nemynm opened the pull request:
Adding support for ANSI X9.63 Key Derivation Function (ANSI-X9.63-KDF).

ANSI-X9.63-KDF is defined in [SEC.1](https://www.secg.org/sec1-v2.pdf) and is the recommended KDF to be used for CMS with ECC - as per [RFC 5753](https://datatracker.ietf.org/doc/html/rfc5753).

This would close issue https://github.com/RustCrypto/KDFs/issues/101.

**Overview:**
This pull request:
 - adds `ansi-x963-kdf` crate with:
    - `ansi_x963_kdf::derive_key_into` function
    - tests for SHA254, SHA256 and SHA512 hash functions
 - updates repository `README.md` with `ansi-x963-kdf`

**Notes:**

- As concat-kdf and ANSI-X9.63-KDF are close and essentially differ in the order in which the data is concatenated, most of the code for this ansi-x963-kdf implementation has been adapted from concat-kdf.
- I did not find public test vectors, so the test cases have been generated using a Java-based implementation of ANSI-X9.63-KDF
- I added one "unsafe" test case to test the input overflow case. As I am not an "unsafe" specialist, carefull review of this one is particularly appreciated.

newpavlov commented on ansi-x963-kdf/src/lib.rs:
```diff
@@ -0,0 +1,121 @@
+    const OVERFLOW_IS_POSSIBLE: bool = core::mem::size_of::<usize>() > 4;
+
+    // 2. Check that keydatalen < hashlen × (2^32 − 1)
+    if OVERFLOW_IS_POSSIBLE && (key.len() >= D::OutputSize::USIZE * (u32::MAX as usize)) {
```
I think this check can be simplified to `key.len() as u64 >= D::OutputSize::U64 * (u32::MAX as u64)`. Compiler is able to properly optimize it out for 32 bit targets: https://rust.godbolt.org/z/rT55T7M7h

nemynm replied:
Addressed in 8322257c67eecbc0aedd9d92d045c8d6e37f7250

newpavlov replied:
I meant that `OVERFLOW_IS_POSSIBLE` constant is not needed and can be removed.

nemynm replied:
Addressed in 400e73db30c2e857416633b3991a86a82b45b6e3

newpavlov commented on ansi-x963-kdf/src/lib.rs:
```diff
@@ -0,0 +1,121 @@
+    // 4. For i = 1 to keydatalen/hashlen,
+    for chunk in key.chunks_mut(D::OutputSize::USIZE) {
+        // 4.1 Compute Ki = Hash(Z ‖ Counter ‖ [SharedInfo]) using the selected hash function
+        Update::update(&mut digest, secret);
```
Why are you not using `digest.update(secret)`? I don't think you need to import the `Update` trait which causes the conflict.

newpavlov replied:
Ah, this is the reason! I forgot about it.

newpavlov commented on ansi-x963-kdf/tests/tests.rs:
```diff
@@ -0,0 +1,245 @@
+            secret: &hex!("00"),
+            other_info: &[],
+            expected_key: &hex!(
+                "4a6ebc83b8e2b19eea640500be6bcffdddaa07b8b2f81f2c533940e4e6ad6cfd
```
Please link provenance of the used test vectors.

nemynm replied:
As I mentioned, I generated these myself using a Java-based implementation - with a few python spot checks. Is that acceptable?
Otherwise I just saw that pyca/cryptography has some test vectors and the license is probably compatible (both LICENSE.APACHE and LICENSE.BSD)? Would that work better maybe?
Let me know what you prefer, thanks.

newpavlov replied:
Yeah, just link the used Java library and write that the vectors were generated by it.

nemynm replied:
Thanks, addressed in 7053d879a43c4be34d5b5ab73cc8c5637fab41fe

newpavlov reviewed (commented):
I have several comments, but otherwise looks good!

Please also add a CI job to test this crate (see the  `.github/workflows` folder).

nemynm commented on ansi-x963-kdf/src/lib.rs:
```diff
@@ -0,0 +1,121 @@
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl ::std::error::Error for Error {}
```
Adressed in 6610e14c1fdb15430b9260a6e686cdeeab742ca3
I still need to fix the failing clippy job at the workspace level now

nemynm commented on ansi-x963-kdf/src/lib.rs:
````diff
@@ -0,0 +1,121 @@
+/// ```rust
+/// let mut key = [0u8; 42];
+/// ansi_x963_kdf::derive_key_into::<sha2::Sha256>(b"top-secret", b"info", &mut key).unwrap();
+/// ```
````
Done in bdd842b6d6a405e8f38e8675203c12e558bde72b

nemynm commented on ansi-x963-kdf/src/lib.rs:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    // 1. Check if |Z| + |SharedInfo| + 4 >= hashmaxlen
+    if secret.len() + shared_info.len() + 4 >= D::OutputSize::USIZE * (u32::MAX as usize) {
```
Adressed in 7053d879a43c4be34d5b5ab73cc8c5637fab41fe
As for:
```
Also, strictly speaking, hashmaxlen depends on hash function algorithm.
```
Does not `D::OutputSize::U64` returns that hash length depending on the hash function considered (the D Generic)?

newpavlov replied:
IIUC `hashmaxlen` here is about maximum size of input data which can be processed by a hash function. Currently we do not expose this value, see #1590.

nemynm replied:
Ah yeah sorry, I got the names mixed, I believe you are right and that it is not a good approximation of hashmaxlen.
As per Sec.1:
>hashmaxlen denote the maximum length in octets of messages that can be hashed using Hash.

I had a quick look at a draft for `ISO/IEC 18033 - by Victor Shoup`, which defines KDF2, and they have the same check:
>a positive integer Hash.MaxInputLen that denotes the maximum length hash input

If it is not exposed, what would be your prefered methods of handling this? From the top of my head I can think of:
- set the hardcoded value to  2^64 - (NIST does seem to mention `2^64` for `SHA1`, `SHA224` and `SHA256`, and `2^128` for for `SHA-384`, `SHA-512`, `SHA-512/224` and `SHA-512/256`)
- inactivate that check as impractical - I think there is no such check for `concat-kdf`, although the linked reference seems to mention it.

In both case we could add a note in the function doc.

newpavlov replied:
The current code is fine, but it's worth to add a comment stating that it's just an approximation. Maybe in future we will properly expose max input len value for hashes.

nemynm replied:
Ok, done in 400e73db30c2e857416633b3991a86a82b45b6e3

newpavlov commented:
You can update the Clippy version to 1.81 in the workspace.yml config to fix the failing CI.

nemynm commented:
> I have several comments, but otherwise looks good!
> 
> Please also add a CI job to test this crate (see the `.github/workflows` folder).

Thanks for reviewing! I tried and addressed your comments. The MSRV bump does trigger a failing clippy at workspace level though.

newpavlov commented:
Thank you!

nemynm commented:
Thanks for the review and the tweaks!
- you probably have noted it, but just in case, on README.md, it seems that the badge MSRV 1.81 and crate on crates.io are not-rendering/missing.
- I think some of these changes could be applied to concat-kdf (e.g. doctests, MSRV, std vs alloc)

In the meantime, I'll close the issue.

newpavlov commented:
>the badge MSRV 1.81 and crate on crates.io are not-rendering/missing.

The MSRV badge (rustc 1.81+) renders fine for me. Docs and crates.io badges require publishing to crates.io, which we will do after releasing digest v0.11.

>I think some of these changes could be applied to concat-kdf (e.g. doctests, MSRV, std vs alloc)

Yes. I plan to do it a bit later.
